NeuroID/NeuroIDClass/Extensions/NIDUser.swift

```diff
@@ -6,156 +6,169 @@
 //
 
 import Foundation
-
 public extension NeuroID {
+    
+    struct SessionIDOriginalResult{
+        let origin: String
+        let originCode: String
+        let idValue: String
+        let idType: UserIDTypes
+    }
+    
     internal static func validateUserID(_ userId: String) -> Bool {
         // user ids must be from 3 to 100 ascii alhpa numeric characters and can include `.`, `-`, and `_`
         do {
             let expression = try NSRegularExpression(pattern: "^[a-zA-Z0-9-_.]{3,100}$", options: NSRegularExpression.Options(rawValue: 0))
             let result = expression.matches(in: userId, options: NSRegularExpression.MatchingOptions(rawValue: 0), range: NSMakeRange(0, userId.count))
             if result.count != 1 {
                 NIDLog.e(NIDError.invalidUserID.rawValue)
-                // If Validation fails send origin event
-                if CURRENT_ORIGIN == nil {
-                    CURRENT_ORIGIN = SessionOrigin.NID_ORIGIN_CUSTOMER_SET.rawValue
-                }
-                CURRENT_ORIGIN_CODE = SessionOrigin.NID_ORIGIN_CODE_FAIL.rawValue
-
-                sendOriginEvent(origin: CURRENT_ORIGIN!, originCode: CURRENT_ORIGIN_CODE!, originSessionID: userId)
-
                 return false
             }
         } catch {
             NIDLog.e(NIDError.invalidUserID.rawValue)
-            // Redundant check to ensure CURRENT_ORIGIN is never unsafely accessed
-            if CURRENT_ORIGIN == nil {
-                CURRENT_ORIGIN = SessionOrigin.NID_ORIGIN_CUSTOMER_SET.rawValue
-            }
-            CURRENT_ORIGIN_CODE = SessionOrigin.NID_ORIGIN_CODE_FAIL.rawValue
-
-            sendOriginEvent(origin: CURRENT_ORIGIN!, originCode: CURRENT_ORIGIN_CODE!, originSessionID: userId)
             return false
         }
-
         return true
     }
-
-    internal static func setGenericUserID(userId: String, type: UserIDTypes, completion: (Bool) -> Bool) -> Bool {
-        let validID = NeuroID.validateUserID(userId)
-        if !validID { return completion(validID) }
-
-        NIDLog.d(tag: "\(type)", "\(userId)")
-
-        let setUserEvent = NIDEvent(
-            sessionEvent: type == .userID ?
+    
+    internal static func setGenericUserID(type: UserIDTypes, genericUserID: String, userGenerated: Bool = true ) -> Bool{
+        let validID = validateUserID(genericUserID)
+        
+        let originRes = getOriginResult(idValue: genericUserID, validID: validID, userGenerated: userGenerated, idType: type)
+        sendOriginEvent(originResult: originRes)
+        
+        if(!validID){return false}
+        
+        NIDLog.d(tag: "\(type)", "\(genericUserID)")
+        //    Queue user id event to be sent
+        var setUserEvent: NIDEvent
+        if (type == .attemptedLogin){
+            setUserEvent = NIDEvent(uid: genericUserID)
```

Comment:
this doesn't have the event type does it? Or because of the constructor it knows its an attempted login

Reply:
That's correct. If no Event Name is provided with uid, it defaults to the attempted login init.

Reply:
https://github.com/Neuro-ID/neuroid-ios-sdk/blob/main/NeuroID/NIDEvent.swift#L467